ambry-mysql/src/main/java/com/github/ambry/mysql/BatchUpdater.java

```diff
@@ -64,29 +64,32 @@
 
   /**
    * Constructor to instantiate a {@link BatchUpdater}.
-   * @param dataAccessor The {@link MySqlDataAccessor}.
+   * @param connection The {@link Connection}.
+   * @param metrics The {@link MySqlMetrics}.
    * @param sql The sql command to prepare the statement.
    * @param tableName The name of the table that update statements are targeting at.
    * @param maxBatchSize The max batch size.
    * @throws SQLException
    */
-  public BatchUpdater(MySqlDataAccessor dataAccessor, String sql, String tableName, int maxBatchSize)
+  public BatchUpdater(Connection connection, MySqlMetrics metrics, String sql, String tableName, int maxBatchSize)
       throws SQLException {
-    this.dataAccessor = Objects.requireNonNull(dataAccessor, "MySqlDataAccessor is empty");
+    this.connection = Objects.requireNonNull(connection, "Connection is empty");
+    this.metrics = Objects.requireNonNull(metrics, "MySqlMetrics is empty");
     this.tableName = Objects.requireNonNull(tableName, "TableName is empty");
     if (maxBatchSize < 0) {
       throw new IllegalArgumentException("MaxBatchSize is not valid: " + maxBatchSize);
     }
     this.maxBatchSize = maxBatchSize;
     try {
       // Calling getPreparedStatement first, since it will setup connection if there is none
-      statement = dataAccessor.getPreparedStatement(sql, true);
+      statement = connection.prepareStatement(sql);
       statement.clearBatch();
-      autoCommit = dataAccessor.getAutoCommmit();
-      dataAccessor.setAutoCommit(false);
+      autoCommit = connection.getAutoCommit();
+      connection.setAutoCommit(false);
     } catch (SQLException e) {
-      dataAccessor.onException(e, BatchUpdate);
+      metrics.batchUpdateFailureCount.inc();
       logger.error("Failed to prepare for batch insert on {}", tableName, e);
+      connection.close();
```

Comment:
we may want to close the prepared statement too if non-null since get/setAutoCommit somehow fails.

Or, does closing the connection automatically release any prepared statement resources?